[ coco ] 미션 4 : 모든 기물 배치하기

choigiseong opened the pull request:
안녕하세요. 백엔드반의 coco라고 합니다.
부족한 코드를 봐주셔서 감사드립니다.

점은 class를 나타냈고 숫자는 메소드를 나타냈습니다.

* ### pieces/Pawn class
원래 PawnEnum으로 있던 color들을 상수로 재정의했습니다.
반면 PawnEnum안에 있던  representation은 새로운 enum으로 재정의해서 사용을 했습니다.

* ### pieces/Representation
각 기물에 대해 구분문자를 가지고 있습니다.
getter를 통해 값에 접근할 수 있습니다.


* ### chess/Board class
1. whiteOtherPieces와 blackOtherPieces를 정의했습니다.
pawn 이외의 piece를 저장하기 위해서입니다.
이로인해 piece를 넣을 때는 addPawn() 과 addOtherPiece()를 사용합니다.
둘은 다른 piece를 받기 때문에 처음에 검증을 진행합니다. 코드가 더러워 보이는데 다른 방법이 생각나지 않아 작성했습니다.

2. clearPieces 메소드
리스트가 늘어남에 따라 코드량이 증가해서 메소드로 분리를 했습니다.  

3. checkColorType 메소드
Piece가 enum타입의 color를 가지고 있다가 String 형태로 된 color로 변경하다 보니
다른 색상이 들어올 수 있다고 생각했습니다. 그래서 색상을 구분해 String을 반환합니다.  


* ### chess/BoardTest class
한 개였던 pawns를 2개의 리스트(white, black)로 분리해서 많은 메소드들이 생성되었습니다.

* ### Printer/PrintChess class
1. printBoard 메소드
   새로만든 StringUtils을 사용합니다. 처음부터 상수에 appendNewLine()메소드를 적용합니다.

긴 글 읽어주셔서 감사드립니다!

choigiseong committed:
StringUtils class 생성과 test

클래스를 생성하고 기본적인 동작을 test했습니다.
같은 기능이라 한번에 커밋합니다.

choigiseong committed:
PrintChess class의 문자열에 appendNewLine()을 사용

상수인 BLANK와 printBoard() 내부의 ...pawns문자열의
끝에 개행문자를 삭제하고 메소드를 이용해
개행시켜준다.
개행문자는 운영체제마다 다를 수 있기 때문에

choigiseong committed:
Pawn class를 Piece class로 이름 변경

class 이름 변경으로 때문에 내부적으로 메소드명과
매개변수명이 변경되었습니다.

choigiseong committed:
PawnEnum의 이름을 Color로 변경

구분문자가 늘어남에 따라 color와 representation을 분리하기위해
먼저 이름 변경했습니다.

choigiseong committed:
Color에서 representation 분리

새로운 enum으로 representation 을 생성했습니다.
내부에는 앞으로 기물별로 다른 구분문자가 들어갈 예정입니다.

choigiseong committed:
Piece의 기본 생성자 삭제

팩토리 메소드를 사용하기 위해 기본 생성자를 삭제하였습니다.

choigiseong committed:
Color enum을 삭제하고 Piece에  Color 상수를 추가

enum이 white,black만 가지고 있어 이점이 없다고 판단하였습니다.
그래서 piece의 상수로 전환 시켰습니다.

choigiseong committed:
Representation enum에 다른 기물들 추가

다른 기물들을 추가했다. 구분문자는 서로 다르다.

choigiseong committed:
각종 기물 팩토리 메소드 추가

기본생성자를 사용하는 팩토리 메소드를 생성했다.

choigiseong committed:
기물 생성 test 작성

기물들이 원하는데로 생성되는지 검사합니다.

choigiseong committed:
폰 이외의 기물 넣기 위해 리스트 생성

whiteOtherPieces와 blackOtherPieces 리스트를 생성했다.
또 addOtherPiece 메소드를 생성했는데 폰 이외의 piece를 받기 위해서 이다.

choigiseong committed:
getPawnsSize 메소드 재정의

pawn들의 크기만이 아닌 모든 기물들의 크기도 return 해준다.

choigiseong committed:
initialize 메소드 수정, 초기화시 모든 기물을 배치

pawn말고 다른 기물들도 알맞게 배치를 한다.

choigiseong committed:
폰 이외의 기물 출력

폰 이외의 기물을 얻을 수 있는 getOtherPiecesResult 메소드를 정의했고
이를 printBoard에도 적용했다.

choigiseong committed:
addOtherPiece 테스트

pawn 이외의 piece를 넣으면 들어가지 않는다.

choigiseong committed:
isBlack, isWhite 메소드 작성과 테스트

알맞은 색상이 들어가는지 테스트를 했다.
print 테스트는 삭제 조치했다. 내용은 initialize 메소드 안으로 옮겼다.

choigiseong committed:
addPawn()과 addOtherPiece()에 예외처리

각기 다른 piece를 다른 리스트에 넣기 때문에
예외처리가 필요해 보여 추가했다.

choigiseong committed:
clearPieces() 추가 와 checkColorType() 예외처리 추가

piece를 담는 리스트가 늘어남에 따라 clear()할 list가 많아져
함수로 추출해 사용했습니다.
또, get...Result()에서 알 수 없는 색상이 들어왔을 때 예외를 발생 시켜줍니다.

choigiseong committed:
isBlack(), isWhite() 리팩토링

테스트 코드도 두 색을 따로 테스트하도록 나눴습니다.

choigiseong committed:
assertThat 문법 수정

앞이 실제값, 뒤가 기댓값으로 재수정 했습니다.

SANGCO commented on src/main/java/net/coco/Printer/PrintChess.java:
```diff
@@ -1,24 +1,28 @@
 package net.coco.Printer;
```
패키지가 대문자로 시작하고 있군요!

choigiseong replied:
수정 완료했습니다!

SANGCO commented on src/main/java/net/coco/pieces/Representation.java:
```diff
@@ -0,0 +1,26 @@
+package net.coco.pieces;
+
+public enum Representation {
```
현재는 'p', 'n' 뭐 이런 데이터만 있지만
향후에 `enum`에 데이터가 추가 된다면
Representation 이라는 이름에 대해서는 한번 고민해볼 필요가 있을거 같네요.

choigiseong replied:
음... 마땅한 이름이 떠오르진 않지만 다음 미션에서 수정하도록 하겠습니다!

SANGCO commented on src/main/java/net/coco/pieces/Representation.java:
```diff
@@ -0,0 +1,26 @@
+package net.coco.pieces;
+
+public enum Representation {
+    WHITE_PAWN('p'),
+    BLACK_PAWN('P'),
+    WHITE_KNIGHT('n'),
+    BLACK_KNIGHT('N'),
+    WHITE_ROOK('r'),
+    BLACK_ROOK('R'),
+    WHITE_BISHOP('b'),
+    BLACK_BISHOP('B'),
+    WHITE_QUEEN('q'),
+    BLACK_QUEEN('Q'),
+    WHITE_KING('k'),
+    BLACK_KING('K');
```
색깔을 나타내는 상수들이 있으니 여기서는 색깔에 관한걸 빼보는건 어떨까요?
'p'를 블랙이 달라고 하면 대문자로 변환해서 'P'로 줄 수도 있지 않을까요?

choigiseong replied:
오! 이런 방법을 생각하지 못했네요. 이 방법을 빨리 알았다면 색상을 붙이지 않았을 건데 감사합니다!

SANGCO commented on src/main/java/net/coco/utils/StringUtils.java:
```diff
@@ -0,0 +1,13 @@
+    }
+
+    public static String appendNewLine(String message) {
+        return message+NEWLINE;
```
`message + NEWLINE;`
사이에 공백을 넣어주세요.

choigiseong replied:
수정 완료했습니다!

SANGCO commented on src/main/java/net/coco/Printer/PrintChess.java:
```diff
@@ -1,24 +1,28 @@
 
 public class PrintChess {
-    private static final String BLANK = "********\n";
+    private static final String BLANK = StringUtils.appendNewLine("********");
```
요게 클래스 변수일 필요가 있을까요?

ChoiGiSung replied:
이 부분에 대해서도 잠깐 고민했었는데, 절대 변하지 않을 거 같아서 남겨놨었습니다.
그럼 299a3afb0588058298bf30111f5e5a9267c07d40 이 부분에 대해선 어떻게 생각하시나요?
이전 리뷰어분께선 start나 end를 상수로 빼는 게 좋다고 하셔서 빼서 사용했습니다. 
제 생각은 start와 end가 변하지 않고 의미있는 값이라 생각해 동의하고 시행했습니다.

현재와 같은 경우는 의미를 호소하는 느낌이 더 떨어져 로컬변수로 변경하는 게 좋을까요?
미래를 생각해 (piece들이 움직일 경우를 위해) 로컬변수로 만들어 변경을 가능하게 하는게 좋을까요?

choigiseong replied:
음.. 이 부분에 대해 잠깐 고민을 했었습니다. 처음 든 생각은 "BLANK는 절대 변하지 않을 거야".

이런 생각이었는데 이번에 다시 든 생각은  
체스말의 이동을 print해 줘야 하니까 바뀔 수 있겠다는 쪽으로 생각이 기울었습니다.
로컬변수로 변경을 하고 후에 변경이 가능해, 기물의 이동을 표시할 수 있게 하겠습니다! 감사합니다!

SANGCO commented on src/main/java/net/coco/Printer/PrintChess.java:
```diff
@@ -1,24 +1,28 @@
-    private static final String BLANK = "********\n";
+    private static final String BLANK = StringUtils.appendNewLine("********");
 
     public static void printBoard(Board board) {
```
`void` 말고 `String`을 리턴하게 수정하고 이 메서드에 대한 테스트 코드를 만들어 보는건 어떨까요?
(체스판에 올라 가야할 피스들이 잘 올라가 있는지 확인하는 테스트)

choigiseong replied:
테스트를 만들어 테스트 해봤습니다. 감사합니다!

SANGCO commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,59 +1,122 @@
+    public void addPawn(Piece piece) {
+        char getRepresentation = piece.getRepresentation();
+        if (!(getRepresentation == Representation.WHITE_PAWN.getRepresentation() || getRepresentation == Representation.BLACK_PAWN.getRepresentation())) {
+            System.out.println("pawn 말고 다른 piece는 넣지 말아주세요.");
+            return;
+        }
+        if (piece.getColor().equals(Piece.WHITE))
+            whitePawns.add(piece);
         else
-            blackPawns.add(pawn);
+            blackPawns.add(piece);
 
     }
 
-    public int getPawnsSize() {
-        return whitePawns.size() + blackPawns.size();
+    public void addOtherPiece(Piece piece) {
+        char getRepresentation = piece.getRepresentation();
+        if (getRepresentation == Representation.WHITE_PAWN.getRepresentation() || getRepresentation == Representation.BLACK_PAWN.getRepresentation()) {
+            System.out.println("pawn은 넣지 말아주세요.");
+            return;
+        }
+        if (piece.getColor().equals(Piece.WHITE))
+            whiteOtherPieces.add(piece);
+        else
+            blackOtherPieces.add(piece);
     }
```
흰색과 검은색을 하나의 메서드에서 만들려고 하니 코드가 지저분해 지는거 같아요.
메소드를 분리해 보는건 어떨까요?

SANGCO replied:
```java
 char getRepresentation = piece.getRepresentation();
 if (!(getRepresentation == Representation.WHITE_PAWN.getRepresentation() || getRepresentation == Representation.BLACK_PAWN.getRepresentation())) {
```
이 부분도 Piece에 `isPawn()` 같은 메서드를 하나 만들면 코드가 좀더 깔끔해 지지 않을까요?

choigiseong replied:
879aace4b212cfc41cd89f3cae07a42b4387ed10 class를 정의해서 piece들의 검증을 책임지는 class를 작성했습니다.
이곳에서 색상, piece의 종류를 검사하게 했습니다.

메소드 분리에 대한 생각은 이렇습니다.
현재 addPawn()과 addOtherPiece()메소드를 사용하는 중인데 
여기서 메소드 분리를 하면 4개의 메소드로 분리가 돼서 더 지저분하지 않을까? 이렇게 생각을 했습니다.
또 저 2개의 메소드가 지는 책임이 "piece를 넣는 역할" 이렇게 생각해서 굳이 white따로 black따로 메소드를 생성하지 않고 메소드 안에 분기를 만들어 진행했었습니다.

SANGCO replied:
> 여기서 메소드 분리를 하면 4개의 메소드로 분리가 돼서 더 지저분하지 않을까? 이렇게 생각을 했습니다.

> 저 2개의 메소드가 지는 책임이 "piece를 넣는 역할" 이렇게 생각해서

단순히 메서드의 숫자가 늘어난다고 그 코드가 지저분하다고 할 수 있을까요?
메서드의 역할이 "piece를 넣는 역할" 이라면 addPawn() 메서드와 addOtherPiece()를 합치면 어떨거 같으세요?
"piece를 넣는 역할"을 하는 메서드 딱 하나
깔끔해 질거 같은가요?

우리가 지향하는 클린코드는 하나의 메서드가 한가지 일만 하게 만드는 거라고 생각해요.
저는 if, else를 발견하면 메서드를 분리할 수 있지 않을까 하고 고민해 본답니다.

우리가 코딩을 하다 보면 보통 하나의 메서드가 점점 길어 지는데요.
코드스쿼드 과정에서 미션을 수행하면서 의도적으로 메서드를 쪼갤 수 있을 때 까지
잘게 분리 시켜보는 연습을 해보는건 어떨까요?

메서드를 4개로 분리 시키면 또 그 메서드들 사이에 비슷하게 생긴 코드가 나올 수도 있고
그러면 그걸 또 어떻게 분리 시킬지 고민하고
이런식으로 연습을 하다보면 하나의 메서드가 하나의 일만하게 되고
메서드 이름만 봐도 어떤 역할을 하는 메서드인지 딱 알 수 있게
코드가 변해가는걸 느끼실 수 있을거 같아요.

아 이런 피드백은 저한테 좀 어렵네요ㅋㅋ
우선 따블 체크를 해주시는 호눅스를 믿고 피드백 남깁니다.
coco님 다른 의견 있으시면 계속 댓글 남겨주세요ㅋ

choigiseong replied:
답변 감사드립니다!
> 단순히 메서드의 숫자가 늘어난다고 그 코드가 지저분하다고 할 수 있을까요?

이 말씀이 다시 한번 생각하게 했습니다.  메소드가 늘어남은 클린코드와 별개로 그 클래스가 지는 책임이 많다는 말이 떠올랐습니다.  그래서 최대한 다른 클래스로 책임을 분산시킬 방법을 찾겠습니다. 아니면 구조적인 문제를 다시 생각해 봐야겠네요.
또 메소드가 늘어나 지저분하다면 `add...()` 메소드를 하나로 합쳤어야 했었네요.

그래서 메소드 분리를 진행해 봤습니다. 
19f3f68 , af682c6 감사합니다!

honux77 replied:
답변에 커밋 참조를 달 수 있군요. 신기해요 ㅎㅎ

SANGCO commented on src/main/java/net/coco/pieces/Piece.java:
```diff
@@ -0,0 +1,91 @@
+    public static final String BLACK = "black";
+
+    private final String color;
+    private final char representation;
+
+    private Piece(String color, char representation) {
+        this.color = color;
+        this.representation = representation;
+    }
```
만드신 `enum`을 활용 할 수 있지 않을까요?

choigiseong replied:
활용하기 위해 enum을 담는 필드를 하나 추가해서 사용했습니다.
PieceType enum을 넘겨주면 representation을 꺼내서 사용하게 수정하였습니다.

SANGCO commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,59 +1,122 @@
                 .collect(Collectors.joining());
     }
 
+    private void addBlackOtherPieceToPieces() {
+        addOtherPiece(Piece.createBlackRook());
+        addOtherPiece(Piece.createBlackKnight());
+        addOtherPiece(Piece.createBlackBishop());
+        addOtherPiece(Piece.createBlackQueen());
+        addOtherPiece(Piece.createBlackKing());
+        addOtherPiece(Piece.createBlackBishop());
+        addOtherPiece(Piece.createBlackKnight());
+        addOtherPiece(Piece.createBlackRook());
+    }
+
+    private void addWhiteOtherPieceToPieces() {
+        addOtherPiece(Piece.createWhiteRook());
+        addOtherPiece(Piece.createWhiteKnight());
+        addOtherPiece(Piece.createWhiteBishop());
+        addOtherPiece(Piece.createWhiteQueen());
+        addOtherPiece(Piece.createWhiteKing());
+        addOtherPiece(Piece.createWhiteBishop());
+        addOtherPiece(Piece.createWhiteKnight());
+        addOtherPiece(Piece.createWhiteRook());
+    }
```
`clearPieces()` 메서드 밑으로 이동하면 좋을거 같네요.
(public 메서드가 사용하는 private 메서드는 public 메서드 밑으로)

choigiseong replied:
이 부분은 너무 길고 굳이 확인하지 않아도 된다고 생각했었습니다.
그래서 규칙은 알고 있지만 맨 밑으로 보내버렸습니다. 
일단 `clearPiece()` 밑으로 이동시켰습니다.

SANGCO commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,59 +1,122 @@
+        checkColorType(color);
+        if (color.equals(Piece.WHITE))
+            return getPiecesToString(whitePawns);
         else
-            return getPawnsToString(blackPawns);
+            return getPiecesToString(blackPawns);
     }
 
-    private String getPawnsToString(List<Pawn> Pawns) {
-        return Pawns.stream()
-                .map(pawn -> String.valueOf(pawn.getRepresentation()))
+    public String getOtherPiecesResult(String color) {
+        checkColorType(color);
+        if (color.equals(Piece.WHITE))
+            return getPiecesToString(whiteOtherPieces);
+        else
+            return getPiecesToString(blackOtherPieces);
+    }
+
+    private void checkColorType(String color) {
+        if (!(color.equals(Piece.WHITE) || color.equals(Piece.BLACK)))
+            throw new IllegalArgumentException("white나 black을 넣어주세요");
+    }
```
이 부분도 `getBlackPawnsResult(), getWhitePawnsResult()` 이렇게 가는게
더 코드가 깔끔 해지지 않을까요?

choigiseong replied:
현재 board는 색상별로, piece별로 (pawn과 otherPiece) 리스트가 존재합니다. 
그래서 리스트별로 내용물을 얻기위해 `getPawnsResult()`와 `getOtherPiecesResult() `메소드를 사용하고 있습니다.
말씀대로 `getBlackPawnsResult(), getWhitePawnsResult()`  2개로 나뉘면 
현재 구조를 변경하지 않는 이상 저 메소드 내부에서 또 분기점으로 pawn인지 otherPiece인지 판단할 거 같습니다.
더 좋은 구조가 있을까요?

SANGCO replied:
이 부분도 위에 댓글 남긴 부분하고 동일한 부분이네요~

SANGCO reviewed (changes requested):
coco님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

choigiseong committed:
패키지명 변경

소문자로 변경했다.

choigiseong committed:
Representation enum에서 색상구분을 삭제

색별로 있던 것을 반환할 때 대문자를 반환하는 것으로 변경해서
코드량이 줄었다.

choigiseong committed:
공백 추가

코드간에 공백 추가

choigiseong committed:
Representation enum의 이름 변경과 getter 변경

원래 색상별로 돌려주다가 getRresentation() 메소드를 정의해서
한번에 돌려주도록 재정의했다.
이로인해 다른 부분에서 변경사항들이 발생해서 변경했다.

choigiseong committed:
Piece class의 필드변경 및 기본생성자 변경

enum을 활용했다. piece는 pieceTpye을 필드로 가지고 있고
이것을 이용해 representation을 얻는다.

choigiseong committed:
PieceValid class 정의

piece를 검증하는 메소드를 작성합니다.
board class가 길어짐에 따라 책임을 분리했습니다.

choigiseong committed:
Board class에서 검증하던 내용을 PieceValid로 책임 전가

검증에 관한 것은 PieceValid가 실행하게 했습니다.
리스트에 알맞은 piece가 들어오는지 검사합니다.

choigiseong committed:
private 메소드를 사용하는 public 메소드 밑으로 이동

직접 사용하는 곳인 public메소드 밑으로 이동했습니다.

choigiseong committed:
isPawn() 검증 수정

pawn과 같지 않을 때 넣으면 안된다.

choigiseong committed:
printBoard() 메소드 테스트

리턴값이 없었지만 test를 위해 리턴값이 있게 했습니다.
test를 진행할 수 있게 되었습니다.

choigiseong committed:
add...() 메소드 분리

addPawn()과 addWhiteOtherPiece를 메소드 분리했다.
메소드의 책임을 더 세분화하여 분리했다.
메소드는 2개에서 4개로 분리되었지만 메소드명으로 확실하게
무슨일을 하는지 알 수 있다.

choigiseong committed:
get...Result() 메소드 분리

4개의 메소드로 분리를 했습니다.
메소드는 하나의 일만 하게 되었고 메소드명을 보고
쉽게 내부동작을 유추할 수 있습니다.

choigiseong committed:
otherWhitePieces를 일급 컬렉션으로 변경

검증하는 쪽이 더러워 보여서 오늘 공부한 일급컬렉션을 도입해 봤습니다.
일단 적용을 확정한게 아니라 otherWhitePieces만 일급컬렉션으로 만들어
봤는데 어떻게 생각하시나요?

choigiseong reviewed (commented):
1b3a6a1
일급컬렉션을 공부하다가 제 코드에도 도입하면 어떨까? 이렇게 생각해서 
도입해 봤습니다.일단 OtherWhitePieces class만 만들었습니다. 

장점은 board class에서 행했던 검증메소드가 OtherWhitePieces클래스 내부에서 이뤄져 
더 깔끔해 보인다, 이 하나만 있는 거 같습니다. 어떻게 생각하시나요? 일단 add 메소드만 깔끔해 지긴 했는데 
안쪽으로 더 책임을 전가하면 board class가 깔끔해 질 수 있겠다고 저는 생각하고 있습니다.

SANGCO reviewed (approved):
coco님 고생 많으셨습니다~
다음 미션도 화이팅 화이팅

choigiseong committed:
add() 검증 수정

pawn이 아닐 때 실행되게 변경
